### DISCUSSION TITLE
Issue with generating consistent citation key for multiple papers with same author and year

### DISCUSSION THREAD

**Discussion by nishan417** [retorquere/zotero-better-bibtex#2933] [Q&A]
Hi,

TLDR: I need help with changing the criteria by which better bibtex assigns "a","b"... to multiple papers with same author and year.

All of my references are in EndNote and recently I started writing a paper on Overleaf and it only supported Zotero so basically the workaround that I have is that I export references from EndNote using refman (RIS) export and import into Zotero. I sync my zotero account with Overleaf where I use those references in a research paper. I use better bibtex to generate citation keys which makes it really easy to refer to specific papers. I am currently using this citation key formula: auth+'_'+year

Every time my references are updated in Endnote, I manually export it to Zotero (i delete my previous references from Zotero). But, I noticed an issue with how it decides to assign citation keys for two of my papers with same author and year. So, with each import into Zotero, sometimes the first paper becomes "John_2023" and the second one becomes "John_2023a". But other times its reversed (i.e., the first paper becomes "John_2023a".). In overleaf, I have cited the first paper as "John_2023", but that changes on the next import. It's really annoying.

**So I did a little test and found out that whichever paper comes first in the exported file is assigned "John_2023" and the second one on the list becomes "John_2023a". Is there a way to change this criterion? For example, maybe alphabetically order the titles of the paper and always assign "a" to the first one? So that whatever sequence I import it into Zotero, the citation key assignment is always consistent?** 

edit: or maybe add the first word of the title ONLY when there's multiple papers with same author/year? for example: John_2023_enrichment, and John_2023_effects. And if there's only one paper then it would just be "John_2023"

edit2: this is only an issue for me because zotero is an intermediate step for me. I regularly add/remove references in Endnote, and to sync it to Zotero, I basically clean up zotero library and import from endnote again. So I need the disambiguation to be consistent for each paper with same author/year.

thank you for your time.

**Comment by retorquere** [comment]
Right-click the entries in Zotero and select "Better BibTeX" -> "Pin citekeys".

**Reply by retorquere** [reply]
There is also an option to auto-pin your citekeys so you  don't have to do this manually: https://retorque.re/zotero-better-bibtex/installation/preferences/index.html#automatically-pin-citation-key-after

**Reply by nishan417** [reply]
My concern is that if I re-import and replace all references in Zotero again then the pinning can't be used since I'm basically starting over. My issue is that I expected the citations keys to be consistent, but it's not. If the original sequence of the references that I'm importing changes, then the key assignment changes.

Ideally I would want the earlier published paper in the same year to be set as "a" and then "b" for the rest. Or, alphabetically order the titles, and assign "a" to the first and "b" to the second. That way a particular paper will always have the same citation key. But with the current design it seems any of those two papers can be "a" depending on the sequence in the imported text file. It seems the software first assigns a key to the first paper then only checks for duplicate for the rest of the paper and modifies the key of the rest of the papers.

If this is not easily possible to do, then I guess I could simply append the first word of the title to the end and make all references unique.

Thanks.

**Reply by retorquere** [reply]
> It seems the software first assigns a key to the first paper then only checks for duplicate for the rest of the paper and modifies the key of the rest of the papers.

That is correct. There is a way around this, but it's roundabout:

1. Sort the items as you want
2. Add `+ 'x'` to the end of your citekey formula
3. Right-click, BBT, refresh
4. remove the `+ 'x'` from the end of your citekey formula
5. Right-click, BBT, refresh

If there is some way to enter data into EndNote such that it ends up in the RIS `M1` or `M2` fields, you can enter

```
Citation Key: whatever
```

there, and that will stay in place all the way to Overleaf. Otherwise what you want isn't really feasible. I have no control over in what order Zotero notifies BBT an item has changed, and I generate the keys as those notifications come in.

**Reply by nishan417** [reply]
Sounds good. Thanks.

(just noticed you're the creator of better bibtex... thanks a lot for your time)
